docs/extensions/guides/ipc.md

```diff
@@ -0,0 +1,92 @@
+
+This is a fully two-way form of communication.
+Both `LensMainExtension` and `LensRendererExtension` can do this sort of IPC.
+
+### Request Based IPC
+
+This is more like a Remote Procedure Call (RPC).
+With this sort of IPC the caller waits for the result from the other side.
+This is accomplished by returning a `Promise<T>` which needs to be `await`-ed.
+
+This is a unidirectional form of communication.
+Only `LensRendererExtension` can initiate this kind of request, and only `LensMainExtension` and respond this this kind of request.
+
+## Registering IPC Handlers and Listeners
+
+The general terminology is as follows:
+
+- A "handler" is the function that responds to a "Request Based IPC" event.
+- A "listener" is the function that is called when a "Event Based IPC" event is emitted.
+
+To register either a handler or a listener, you should do something like the following:
+
+```typescript
+import { LensMainExtension, Interface, Types } from "@k8slens/extensions";
+
+export class ExampleExtensionMain extends LensMainExtension {
+  onActivate() {
+    this.disposers.push(
+      this.listenIpc({
+        channel: "initialize",
+        listener: this.initializeListener,
+        verifier: this.initializeVerifier,
+      })
+    );
+  }
+
+  initializeListener = (event: Types.IpcMainEvent, uid: string) => {
+    console.log(`starting to initialize: ${uid}`);
+  };
+
+  initializeVerifier = (args: unknown[]): args is [uid: string] => {
+    return args.length === 1 && typeof args[0] === "string";
+  }
+}
+```
+
+If you want to register a "handler" you would call `this.handleIpc(...)` instead.
+
+The `LensExtension.prototype.disopsers` is a list of `() => void`'s (or `Utils.Disposer`'s).
+Those functions are sort of like destructors in that they should clean up some state.
+Some functions return them to indicate that it will clean up the state which was just registered.
+
+As an extension developer you do not need to run them yourself.
+Lens will run them when your extension gets deactivated or uninstalled.
+
+### Note about verification
+
+We require that extension developers provide a verification function when registering a listener or handler.
+This is done as a preventative measure to help separate issues that can happen at runtime.
+While it is possible to use the unary truth function, this is highly discouraged.
+
+The verification function should do some cursory validation on the values send along the channel.
+Your handler or listener will not be called if it fails this validation.
+Instead an error or log message will occur.
+This should help with debugging because you are notified immediately that there is a mismatch between what you are expecting and what was sent.
+
+## Using IPC
+
+Calling IPC is very simple.
+If you are meaning to do an event based call, merely call `this.sendIpc(<channel>, ...<args>)` from within your extension.
```

Comment:
Is there a restriction to the data types that can be included in `args`? Do they all have to be JSON-serializable types, or can they be things like ES6 `Map` and `Set`, or even `Function` (things that are beyond the scope of JSON, but that I can freely pass around between functions within a normal JS program)?

Reply:
That is probably a good thing to call out. It can be more than JSON-serializable. It has to follow the https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm.

So things like `Map` and `Set` are good but `Function` will throw an error

Reply:
I think so, yes. That should definitely be called out.